Sources/SwiftSyntax/RawSyntaxNodeProtocol.swift

```diff
@@ -0,0 +1,124 @@
+//===------- RawSyntaxNodeProtocol.swift - Typed Raw Syntax Protocol ------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// All typed raw syntax nodes conform to this protocol.
+/// `RawXXXSyntax` is a typed wrappeer of `RawSyntax`.
+@_spi(RawSyntax)
+public protocol RawSyntaxNodeProtocol {
+  /// Returns `true` if `raw` can be cast to this concrete raw syntax type.
+  static func isKindOf(_ raw: RawSyntax) -> Bool
+
+  /// Type erased raw syntax.
+  var raw: RawSyntax { get }
+
+  /// Create the typed raw syntax if `other` can be cast to `Self`
+  init?<T: RawSyntaxNodeProtocol>(_ other: T)
+}
+
+public extension RawSyntaxNodeProtocol {
+  /// Cast to the specified raw syntax type if possible.
+  func `as`<Node: RawSyntaxNodeProtocol>(_: Node.Type) -> Node? {
+    Node(self)
+  }
+
+  /// Check if this instance can be cast to the specified syntax type.
+  func `is`<Node: RawSyntaxNodeProtocol>(_: Node.Type) -> Bool {
+    Node.isKindOf(self.raw)
+  }
+}
+
+/// `RawSyntax` itself conforms to `RawSyntaxNodeProtocol`.
+extension RawSyntax: RawSyntaxNodeProtocol {
+  public static func isKindOf(_ raw: RawSyntax) -> Bool {
+    return true
+  }
+  public var raw: RawSyntax { self }
+
+  init(raw: RawSyntax) {
+    self = raw
+  }
+
+  public init<T: RawSyntaxNodeProtocol>(_ other: T) {
+    self.init(raw: other.raw)
+  }
+}
+
+@_spi(RawSyntax)
+public struct RawTokenSyntax: RawSyntaxNodeProtocol {
+  public static func isKindOf(_ raw: RawSyntax) -> Bool {
+    return raw.kind == .token
+  }
+
+  public var raw: RawSyntax
+
+  init(raw: RawSyntax) {
+    assert(Self.isKindOf(raw))
+    self.raw = raw
+  }
+
+  public init?<Node: RawSyntaxNodeProtocol>(_ other: Node) {
+      guard Self.isKindOf(other.raw) else { return nil }
+    self.init(raw: other.raw)
+  }
+
+  public var tokenKind: RawTokenKind {
+    return raw.rawTokenKind
+  }
+
+  public var tokenText: SyntaxText {
+    return raw.rawTokenText
+  }
+
+  public var presence: SourcePresence {
+    raw.presence
+  }
+
+  public var isMissing: Bool {
+    presence == .missing
+  }
+
+  /// Creates a `TokenSyntax`. `text` and trivia must be managed by the same
+  /// `arena`.
+  public init(
+    kind: RawTokenKind,
+    text: SyntaxText,
+    leadingTriviaPieces: [RawTriviaPiece],
+    trailingTriviaPieces: [RawTriviaPiece],
+    arena: SyntaxArena
+  ) {
+    let totalTriviaCount = leadingTriviaPieces.count + trailingTriviaPieces.count
```

Comment:
We can. However, they can actually be a static text like:
```
RawTokenSyntax(
  tokenKind: .structKeyword, tokenText: "struct",
  leadingTriviaPiecies: [.docLineComment("/// Foo"), .newlines(1)], trailingTriviaPieces: [],
  arena: arena)
```
This is safe because static text are always alive. Do you think this feature is not needed?

Reply:
This was basically referring to your comment in line 90: “`text` and trivia must be managed by the same `arena`.” So I thought whether we could enforce that requirement using assertions.

Reply:
Added the assertions.